src/pudl/transform/epacems.py

```diff
@@ -42,160 +46,243 @@ def harmonize_eia_epa_orispl(
+        # deprecated, but the PyArrow schema stores this data as UTC. See:
+        # https://numpy.org/devdocs/reference/arrays.datetime.html#basic-datetimes
+        .with_columns(
+            operating_datetime_utc=pl.col("op_datetime_naive") - pl.col("utc_offset")
         )
-    # Add the offset from UTC. CEMS data don't have DST, so the offset is always the
-    # same for a given plant. The result is a timezone naive datetime column that
-    # contains values in UTC. Storing timezone info in Numpy datetime64 objects is
-    # deprecated, but the PyArrow schema stores this data as UTC. See:
-    # https://numpy.org/devdocs/reference/arrays.datetime.html#basic-datetimes
-    df["operating_datetime_utc"] = df["op_datetime_naive"] - df["utc_offset"]
-    del (
-        df["op_date"],
-        df["op_hour"],
-        df["op_datetime_naive"],
-        df["utc_offset"],
+        .drop(["op_date", "op_hour", "op_datetime_naive", "utc_offset"])
     )
-    return df
 
 
-def _load_plant_utc_offset(core_eia__entity_plants: pd.DataFrame) -> pd.DataFrame:
-    """Load the UTC offset each EIA plant.
+def _load_plant_utc_offset(core_eia__entity_plants: pl.DataFrame) -> pl.DataFrame:
+    """Load the UTC offset for each EIA plant.
 
     CEMS times don't change for DST, so we get the UTC offset by using the
     offset for the plants' timezones in January.
 
     Args:
-        pudl_engine: A database connection engine for
-            an existing PUDL DB.
+        core_eia__entity_plants: EIA plants DataFrame.
 
     Returns:
-        Dataframe of applicable timezones taken from the core_eia__entity_plants table.
+        Polars DataFrame of applicable timezones taken from the core_eia__entity_plants
+        table.
     """
-    timezones = core_eia__entity_plants[["plant_id_eia", "timezone"]].copy().dropna()
-    jan1 = datetime.datetime(2011, 1, 1)  # year doesn't matter
-    timezones["utc_offset"] = timezones["timezone"].apply(
-        lambda tz: pytz.timezone(tz).localize(jan1).utcoffset()
+    logger.debug("Creating plant UTC offset DataFrame")
+
+    # Create a mapping of unique timezones to offsets to avoid repeated calculations
+    jan1 = datetime.datetime(2011, 1, 1)
+
+    timezone_offset_map = {
+        tz: pytz.timezone(tz).localize(jan1).utcoffset()
+        for tz in core_eia__entity_plants.get_column("timezone")
+        .drop_nulls()
+        .unique()
+        .to_list()
+    }
+
+    return (
+        core_eia__entity_plants.select(["plant_id_eia", "timezone"])
+        .drop_nulls()
+        .unique()
+        .with_columns(
+            utc_offset=pl.col("timezone").replace(timezone_offset_map, default=None)
+        )
+        .select(["plant_id_eia", "utc_offset"])
     )
-    del timezones["timezone"]
-    return timezones
 
 
-def correct_gross_load_mw(df: pd.DataFrame) -> pd.DataFrame:
-    """Fix values of gross load that are wrong by orders of magnitude.
```

Comment:
For future travelers: this check is now short & clear enough to do inline during `transform_epacems()`